### PR TITLE
Determine random activity based on user selections

### LINKED ISSUE #50: Filter all activities based on user's selections

### PR DESCRIPTION
**Pull request by nicolegooden**
#### What's this PR do?  

+ Makes fetch request for data per activity when the activities are set in App's state
+ Refactors `getActivityData()` so that it can be used in multiple places (fetching when activities are chosen, and fetching when the user chooses to browse our content)
+ Adds link to the Form's submit button so that the user can see the suggested activity (soon to be working)
+ Creates several algorithms for filtering based on each activity's questions
+ Leads `determineRandomActivity()` to return a randomized object - useful for the random activity display

#### Where should the reviewer start? 

Most changes have occurred in `App.js`.  Review the changes in the methods in this file, and let me know if we've missed any edge cases that stick out.

#### How should this be manually tested?  

This can only be tested with `console logs` right now, so you can pull down this version locally, and track the value of `possibleSuggestions` in `filterActivity()` within `App.js` as you simulate user interactions with the form.

#### Any background context you want to provide?  

The `filterActivity()` method is super long - we can refactor it by creating several helper functions to do the same work in a different space.  

#### What are the relevant tickets?  

Closes #50

### CODE REVIEW

File: src/Components/App/App.js [hunk 1]
```diff
@@ -31,25 +31,31 @@ class App extends Component {
...
   }
 
-  getActivityData = async (event) => {
+  getActivityData = async (name) => {
+    if (name.target) {
+      name = name.target.id
+    }
+    if (name.includes('games')) {
+      name = name.replace(/ games/gi, 'Games')
+    }
```
Comment (Atos20) [inline src/Components/App/App.js:40]: I like how readable this is, thanks for improving this method.

File: src/Components/App/App.js [hunk 2]
```diff
@@ -31,25 +31,31 @@ class App extends Component {
...
         promise = await getAllMusic();
       }
-      if (event.target.id === 'podcasts') {
+      if (name === 'podcasts') {
         promise = await getAllPodcasts();
       }
-      this.setState({[event.target.id]: promise})
+      this.setState({[name]: promise})
     } catch (error) {
       console.log(error)
```
Comment (Atos20) [inline src/Components/App/App.js:60]: I think we want to assign the error to a property in the state so we can render appropriate content, such as failed to retrieve for error handling.

**Review by Atos20 — APPROVED** [review]
Thanks for all your hard work

File: src/Components/App/App.js [hunk 3]
```diff
@@ -31,25 +31,31 @@ class App extends Component {
     }
   }
 
-  getActivityData = async (event) => {
+  getActivityData = async (name) => {
+    if (name.target) {
+      name = name.target.id
+    }
+    if (name.includes('games')) {
+      name = name.replace(/ games/gi, 'Games')
```
Comment (BlakeDonn) [inline src/Components/App/App.js:39]: Maybe we can think of a way to not use regex here, but for now it does the trick.

File: src/Components/App/App.js [hunk 4]
```diff
@@ -62,22 +68,74 @@ class App extends Component {
...
+        activity = activity.replace(/ games/gi, 'Games')
+      }
+      let filteredActivities = this.filterActivity(activity, this.state[`${activity}Answers`])
+      filtered.push(filteredActivities)
+      return filtered
+    }, [])
+    let final = allFilteredActivities.flat()
+    let randomNumber = Math.floor(Math.random() *final.length)
+    return final[randomNumber]
+  }
```
Comment (BlakeDonn) [inline src/Components/App/App.js:89]: While it's not the prettiest function, It has some pretty solid logic regarding the filtering of the activities based on user responses.

File: src/Components/App/App.js [hunk 5]
```diff
@@ -62,22 +68,74 @@ class App extends Component {
...
+    let gamesGroup = ['boardGames', 'cardGames']
+    let ageGroup = ['movies', 'music']
+    let possibleSuggestions;
+    if (genreGroup.includes(activity)) {
+      possibleSuggestions = this.state[activity].filter(singleActivity => {
+        if (activity === 'movies') {
+          return singleActivity.genre.some(genre => answers.includes(genre))
+        }
+        return answers.some(genre => genre === singleActivity.genre)
+      })
```
Comment (BlakeDonn) [inline src/Components/App/App.js:103]: While it's not the prettiest function, It has some pretty solid logic regarding the filtering of the activities based on user responses.

I think we can definitely split these up separate functions that are only ever responsible for the specific activity type

File: src/Components/App/App.js [hunk 6]
```diff
@@ -62,22 +68,74 @@ class App extends Component {
...
+    }
+    if (activity === 'movies') {
+        if (answers.includes('too long')) {
+          possibleSuggestions = possibleSuggestions.filter(element => {
+            return +element.runtime < 120
+          })
+        }
+        if (answers.includes('that\'s fine')) {
+          return possibleSuggestions
+        }
```
Comment (BlakeDonn) [inline src/Components/App/App.js:123]: We ran into some issues here regarding not all answers being filled out, which would ultimately break this functionality.

We will definitely need to add in some restrictions that the user must answer all questions at least once before proceeding

File: src/Components/Form/Form.js [hunk 7]
```diff
@@ -19,24 +19,35 @@ export class Form extends Component {
...
+    if (+questionData.allowedResponses && this.state.currentAnswers.find(answer => questionData.choices.includes(answer))) {
+      let choiceToSwap = this.state.currentAnswers.find(answer => questionData.choices.includes(answer))
+      let indexToReplace = this.state.currentAnswers.indexOf(choiceToSwap)
+      let answerToReplace = this.state.currentAnswers
+      answerToReplace.splice(indexToReplace, 1, event.target.textContent)
+      if (answerToReplace) {
+        return this.setState({currentAnswers: answerToReplace})
+      }
+    } else {
+      this.setState({currentAnswers: [...this.state.currentAnswers, event.target.textContent]})
```
Comment (BlakeDonn) [inline src/Components/Form/Form.js:39]: I have noticed some weird functionality on clicking where the current answers doesn't always get updated. I hope this function isn't the reason for that but think its definitely worth considering

**Review by BlakeDonn — APPROVED** [review]
All changes bring our product that much closer to a finalized product! Functionality is pretty much complete with these changes

**Pull request merged**